### PR TITLE
Page Structure + What Is A Custom Device page

### PR DESCRIPTION
**Pull request by cchidean**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-custom-device-handbook/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Updated the index.md file and content for What_Is_A_Custom_Device.md

### Why should this Pull Request be merged?

Adds new and exciting content to the custom devices handbook

### What testing has been done?

Read the Docs build passed: https://niveristand-custom-device-handbook--8.org.readthedocs.build/en/8/What_Is_A_Custom_Device.html

### CODE REVIEW

File: docs/What_Is_A_Custom_Device.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+The project is pre-populated with VIs, LabVIEW Libraries, an XML File, and two build specifications. These resources provide the framework upon which almost all custom devices are built.
+
+VeriStand evolved from NI Dynamic Test Software (NI-DTS). NI-DTS evolved from 3rd party intellectual property (IP) called EASE. The IP made basic provisions for add-on LabVIEW code. This provisions could be considered the first custom device framework on which several “custom devices” were built. If you find a custom device that does not fit the Custom Device Template Tool framework, you may be operating an EASE based custom devices.
+
+For each of the five types of custom devices, you’ll see two VI libraries in the LabVIEW source project: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Name Custom Device.lvlib.*</span>
+
+The following image displays a new custom device project that uses these two libraries.
+![](images/Picture5.jpg)
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:105]: We need to replace this image, with:
image

File: docs/What_Is_A_Custom_Device.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+The developer specifies the type of custom device before running the template tool. The tool generates the LabVIEW Project for the new custom device. The exact resources in the project depend on the type of custom device selected.
+
+The project is pre-populated with VIs, LabVIEW Libraries, an XML File, and two build specifications. These resources provide the framework upon which almost all custom devices are built.
+
+VeriStand evolved from NI Dynamic Test Software (NI-DTS). NI-DTS evolved from 3rd party intellectual property (IP) called EASE. The IP made basic provisions for add-on LabVIEW code. This provisions could be considered the first custom device framework on which several “custom devices” were built. If you find a custom device that does not fit the Custom Device Template Tool framework, you may be operating an EASE based custom devices.
+
+For each of the five types of custom devices, you’ll see two VI libraries in the LabVIEW source project: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Name Custom Device.lvlib.*</span>
+
+The following image displays a new custom device project that uses these two libraries.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:104]:
```suggestion
The following image displays a new custom device template project.
```

File: docs/What_Is_A_Custom_Device.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+The developer specifies the type of custom device before running the template tool. The tool generates the LabVIEW Project for the new custom device. The exact resources in the project depend on the type of custom device selected.
+
+The project is pre-populated with VIs, LabVIEW Libraries, an XML File, and two build specifications. These resources provide the framework upon which almost all custom devices are built.
+
+VeriStand evolved from NI Dynamic Test Software (NI-DTS). NI-DTS evolved from 3rd party intellectual property (IP) called EASE. The IP made basic provisions for add-on LabVIEW code. This provisions could be considered the first custom device framework on which several “custom devices” were built. If you find a custom device that does not fit the Custom Device Template Tool framework, you may be operating an EASE based custom devices.
+
+For each of the five types of custom devices, you’ll see two VI libraries in the LabVIEW source project: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Name Custom Device.lvlib.*</span>
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:102]:
```suggestion
For each of the five types of custom devices, in the LabVIEW source project, you’ll see:
* an _NI VeriStand APIs_ Virtual Folder containing two VI libraries: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Utility Library.lvlib.*</span>;
* a <span style="color:green">*Custom Device Name Engine.lvlib.*</span> VI library;
* a <span style="color:green">*Custom Device Name Shared.lvlib.*</span> VI library;
* a <span style="color:green">*Custom Device Name System Explorer.lvlib.*</span> VI library.
```

File: docs/What_Is_A_Custom_Device.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+The following image displays a new custom device project that uses these two libraries.
+![](images/Picture5.jpg)
+<br />
+
+The Custom Device API library contains most of the type definitions, template VIs and LabVIEW API needed to interact with VeriStand data and timing resources. They allow the VI to behave as a native task in the VeriStand Engine.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:108]:
```suggestion
The Custom Device API library and Custom Device Utility Library contain most of the type definitions, template VIs and LabVIEW API needed to interact with VeriStand data and timing resources. They allow the VI to behave as a native task in the VeriStand Engine.
```

File: docs/What_Is_A_Custom_Device.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+For each of the five types of custom devices, you’ll see two VI libraries in the LabVIEW source project: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Name Custom Device.lvlib.*</span>
+
+The following image displays a new custom device project that uses these two libraries.
+![](images/Picture5.jpg)
+<br />
+
+The Custom Device API library contains most of the type definitions, template VIs and LabVIEW API needed to interact with VeriStand data and timing resources. They allow the VI to behave as a native task in the VeriStand Engine. 
+
+&nbsp;&nbsp;&nbsp; **Note:** Some of these VIs also appear on the LabVIEW palette in **[VeriStand](https://zone.ni.com/reference/en-XX/help/372846M-01/veristandmerge/vs_ni_veristand_vis_pal/)** » **[Custom Device API](https://zone.ni.com/reference/en-XX/help/372846M-01/veristandmerge/vs_custom_device_api_vis_pal/)**.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:110]:
```suggestion
&nbsp;&nbsp;&nbsp; **Note:** Some of these VIs also appear on the LabVIEW palette in **[NI VeriStand](https://zone.ni.com/reference/en-XX/help/372846M-01/veristandmerge/vs_ni_veristand_vis_pal/)** » **[Custom Device API](https://zone.ni.com/reference/en-XX/help/372846M-01/veristandmerge/vs_custom_device_api_vis_pal/)**.
```

File: docs/What_Is_A_Custom_Device.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+The API library contains the custom device’s configuration and real-time engine VIs. These correspond to the configuration and engine VI libraries (or LLBs). The front panel and block diagram of these VIs are populated with objects from the Custom Device API library.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:112]:
```suggestion
The API library contains the custom device’s configuration and real-time engine VIs. These correspond to the configuration and engine VI libraries (or LLBs). The front panel and block diagram of these VIs are populated with objects from the Custom Device API libraries.
```

File: docs/What_Is_A_Custom_Device.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+#### Configuration
+
+The custom device’s configuration defines how the operator adds and configures the custom device through a user interface (UI). The Custom Device Template Tool provides the Initialization VIs for configuration purposes. You can add more VIs during development. 
+
+When a custom device VI’s front panel is presented to the operator in System Explorer, that VI is called a page. Pages are a subset of the VIs that make up a custom device.
+
+
+#### Initialization VI
+
+The Custom Device Template Tool names the initialization VI <span style="color:green">*<Custom Device Name> Initialization VI.vi.*</span> It runs in the background when the custom device is first added to the system definition. The initialization page does not run again unless the operator removes and re-adds the custom device.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:124]:
```suggestion
The niveristand-custom-device-wizard adds the <span style="color:green">*Initialization VI.vi.*</span>. This VI runs in the background when the custom device is first added to the system definition. The initialization page does not run again unless the operator removes and re-adds the custom device.
```

File: docs/What_Is_A_Custom_Device.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+While you may rename certain objects in the custom device’s LabVIEW Project, it’s important to understand the ramifications of doing so. For example, the Initialization VI is referenced by name in the custom device XML file. This file is generated when you first run the Custom Device Template Tool. If you rename the Initialization VI after running the tool, you’ll need to manually change the path to the Initialization VI in the
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:126]:
```suggestion
While you may rename certain objects in the custom device’s LabVIEW Project, it’s important to understand the ramifications of doing so. For example, the Initialization VI is referenced by name in the custom device XML file. This file is generated when you first run the niveristand-custom-device-wizard. If you rename the Initialization VI after running the wizard, you’ll need to manually change the path to the Initialization VI in the
```

File: docs/What_Is_A_Custom_Device.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+custom device XML file.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:127]:
```suggestion
```

File: docs/What_Is_A_Custom_Device.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+The Initialization Page runs each time a new instance of the same custom device is added to the system definition. VeriStand retains state information for each instance of a custom device in the **[System Definition](https://zone.ni.com/reference/en-XX/help/372846M-01/veristand/comp_of_project/)** (.**[nivssdf](https://www.ni.com/documentation/en/veristand/latest/manual/configure-system-definition-file/)**) file. State is defined by the value of each control, indicator, and property of the page. The system definition is human-readable XML, so you can open the file with a text editor. 
+
+&nbsp;&nbsp;&nbsp; **Note:** You can use the **[.NET API](https://www.ni.com/documentation/en/veristand/latest/manual/veristand-net-reference/)** to programmatically modify the system definition.
+
+#### Main Page
+
+The Custom Device Template Tool creates Main Page.vi. After the custom device has been added to the system definition, the main page runs whenever the operator clicks on the custom device’s top-level item in the System Explorer configuration tree. The following image displays the top-level item.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:135]:
```suggestion
The niveristand-custom-device-wizard creates Main Page.vi inside Dynamically Called virtual folder of the System Explorer library. After the custom device has been added to the system definition, the Main Page runs whenever the operator clicks on the custom device’s top-level item in the System Explorer configuration tree. The following image displays the top-level item.
```

File: docs/What_Is_A_Custom_Device.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+![](images/Picture6.png)
+<br />
+
+#### Engine
+
+The Custom Device Template Tool creates the <span style="color:green">*<Custom Device Name> RT Driver.vi.*</span> This defines the behavior of the custom device on the **[execution host](http://zone.ni.com/reference/en-XX/help/372846B-01/veristand/veristand_glossary/#execution_host)**. The RT Driver VI runs on the execution host regardless of the target’s operating system. VeriStand deploys the engine when the operator runs the project from VeriStand or when the system definition is deployed using the VeriStand Execution API.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:142]:
```suggestion
The niveristand-custom-device-wizard creates the <span style="color:green">*&lt;Custom Device Name&gt; RT Driver.vi.*</span> This defines the behavior of the custom device on the **[execution host](http://zone.ni.com/reference/en-XX/help/372846B-01/veristand/veristand_glossary/#execution_host)**. The RT Driver VI runs on the execution host regardless of the target’s operating system. VeriStand deploys the engine when the operator runs the project from VeriStand or when the system definition is deployed using the VeriStand Execution API.
```

File: docs/What_Is_A_Custom_Device.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+VeriStand 2009 did not support the VeriStand Engine on VxWorks operating systems. Starting with VeriStand 2010, if you want to support VxWorks targets such as Compact RIO, you must compile the engine library for VxWorks. PharLap and Windows engines do not require additional compilation.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:144]: We remove this. Outdated. OSes not supported anymore.
```suggestion
```

File: docs/What_Is_A_Custom_Device.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+Each of the five prebuilt custom devices has a different engine VI. Each engine VI executes at a different time with respect to other VeriStand components. The timing requirements of a custom device, and thus the type of device selected, are functions of when the device needs to execute with respect to other VeriStand Engine components.
+
+Not all requirements can be satisfied by one of the five types of prebuilt custom devices. Some custom devices will require multiple engine libraries, such as (to support different real-time operating systems for example). **[NI VeriStand – Set Custom Device Driver VI](https://zone.ni.com/reference/en-XX/help/372846M-01/veristandmerge/vs_set_custom_device_drivers/)** allows you to programmatically change the driver library for a custom device. 
+Some custom devices use the prebuilt template as a launching pad for multiple parallel processes or complex frameworks.
+For more information, refer to **Beyond the Template Frameworks**.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:152]: We should make the **Beyond the Template Frameworks** visible somehow that is should be a hiperlink in the future. Maybe make it Red and add a note (insert hiperlink) or something, to not forget to connect it when the page is ready.

File: docs/What_Is_A_Custom_Device.md [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+Custom code performs any functionality desired by the custom device developer. While the initialization and engine frameworks provide access to VeriStand data and timing resources, you must implement the code to meet specification.
+
+For example, custom code can perform a single A/D conversion on a 3rd party digitizer. The framework provides the means for sending the digitized value to the rest of the VeriStand system so that it can be mapped to a channel or used in a stimulus profile.
+
+#### Custom Device XML
+
+Each custom device has an XML file that contains information used by VeriStand to load, configure, display, deploy and run the device. The basic information includes VI and dependency paths, page names, action items, menu items, and meta data for the various pages that make up the custom device.
+
+The Custom Device Template Tool generates an XML file in the template LabVIEW Project. Any properly formatted XML file will be parsed by VeriStand. After the XML file is created by the Custom Device Template Tool, all updates have to be manual.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:164]:
```suggestion
The niveristand-custom-device-wizard generates an XML file in the template LabVIEW Project. Any properly formatted XML file will be parsed by VeriStand. After the XML file is created by the Custom Device Template Tool, all updates have to be manual.
```

File: docs/What_Is_A_Custom_Device.md [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+The custom device XML file does not automatically synchronize with changes to the LabVIEW project. Also, the file does not automatically deploy. You must modify the XML file in the LabVIEW Project directory when making changes. Building the initialization specification overwrites the XML in the <span style="color:green">*&lt;Common Data&gt;\Custom Devices folder*</span>.
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:166]:
```suggestion
The custom device XML file does not automatically synchronize with changes to the LabVIEW project. Also, the file does not automatically deploy. You must modify the XML file in the LabVIEW Project directory when making changes. Building the initialization specification overwrites the XML in the <span style="color:green">*&lt;Common Data&gt;\Custom Devices*</span> folder.
```

**Review by phutanu — CHANGES_REQUESTED** [review]
Please check my above suggestions. Thanks for your patience. :)

**Comment by WillCawley** [comment]
Hey @phutanu and @cchidean, how close is this PR to being ready to submit? I think it would be easier for me to review the files if I could make my own PR for a tech writer review.

File: docs/What_Is_A_Custom_Device.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+The project is pre-populated with VIs, LabVIEW Libraries, an XML File, and two build specifications. These resources provide the framework upon which almost all custom devices are built.
+
+VeriStand evolved from NI Dynamic Test Software (NI-DTS). NI-DTS evolved from 3rd party intellectual property (IP) called EASE. The IP made basic provisions for add-on LabVIEW code. This provisions could be considered the first custom device framework on which several “custom devices” were built. If you find a custom device that does not fit the niveristand-custom-device-wizard framework, you may be operating an EASE based custom devices.
+
+For each of the five types of custom devices, in the LabVIEW source project, you’ll see:
+* an _NI VeriStand APIs_ virtual folder containing two VI libraries: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Utility Library.lvlib.*</span>;
+* a <span style="color:green">*&lt;Custom Device Name&gt; Engine.lvlib.*</span> VI library;
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:95]:
```suggestion
* a <span style="color:green">*&lt;Custom Device Name&gt; Engine.lvlib*</span> VI library;
```

File: docs/What_Is_A_Custom_Device.md [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+* a <span style="color:green">*&lt;Custom Device Name&gt; Shared.lvlib.*</span> VI library;
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:96]:
```suggestion
* a <span style="color:green">*&lt;Custom Device Name&gt; Shared.lvlib*</span> VI library;
```

File: docs/What_Is_A_Custom_Device.md [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+The developer specifies the type of custom device before running the niveristand-custom-device-wizard. The wizard generates the LabVIEW Project for the new custom device. The exact resources in the project depend on the type of custom device selected.
+
+The project is pre-populated with VIs, LabVIEW Libraries, an XML File, and two build specifications. These resources provide the framework upon which almost all custom devices are built.
+
+VeriStand evolved from NI Dynamic Test Software (NI-DTS). NI-DTS evolved from 3rd party intellectual property (IP) called EASE. The IP made basic provisions for add-on LabVIEW code. This provisions could be considered the first custom device framework on which several “custom devices” were built. If you find a custom device that does not fit the niveristand-custom-device-wizard framework, you may be operating an EASE based custom devices.
+
+For each of the five types of custom devices, in the LabVIEW source project, you’ll see:
+* an _NI VeriStand APIs_ virtual folder containing two VI libraries: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Utility Library.lvlib.*</span>;
```
Comment (phutanu) [inline docs/What_Is_A_Custom_Device.md:94]:
```suggestion
* an _NI VeriStand APIs_ virtual folder containing two VI libraries: <span style="color:green">*Custom Device API.lvlib*</span> and <span style="color:green">*Custom Device Utility Library.lvlib*</span>;
```

**Pull request merged**